File: swift/Sources/CircleAI/PersonalHealth.swift
// PersonalHealth.swift
//
// Port of the Personal.Health vertical from
// src/CircleAI.Personal.Health/PersonalHealthPrimitives.cs and the static
// domain-context constants from PersonalHealthDomainContext.cs:
//   • VitalKind (enum), VitalReading, Allergy, Medication — records
//   • IPersonalHealthBoard        — vitals / allergies / medications
//   • InMemoryPersonalHealthBoard — deterministic in-memory impl
//   • PersonalHealthDomainContext — system-prompt snippet + flags
//
// The Companion-facing wrapper (PersonalHealthCompanionAdapter) is intentionally
// NOT ported (see Healthcare.swift for the rationale).
//
// Porting notes:
//   • `VitalKind` is an `Int`-backed enum so its ordinal (0…7) matches the C#
//     enum's default underlying values; `CaseIterable` mirrors iteration.
//   • `DateTimeOffset` → `Date`; `DateTimeOffset?` (Medication.endedAtUtc) → `Date?`.
//   • `EndMedication` on an unknown med throws → `PersonalHealthError.unknownMedication`.
//   • `ReadSince` orders ascending by time; `Latest` is the newest reading of a
//     kind (or `nil`). `ActiveMedications` = those with no end date, ordered by
//     name.

import Foundation

// MARK: - Enums

/// A kind of vital sign. `Int`-backed so ordinals match the C# enum.
public enum VitalKind: Int, Sendable, Codable, CaseIterable {
    case bloodPressureSystolic = 0
    case bloodPressureDiastolic
    case glucoseMgDl
    case weightKg
    case heartRateBpm
    case temperatureC
    case oxygenPct
    case stepsCount
}

// MARK: - Records

/// A single vital-sign reading.
public struct VitalReading: Sendable, Equatable, Codable {
    /// Which vital was measured.
    public let kind: VitalKind
    /// Measured value.
    public let value: Double
    /// UTC timestamp.
    public let atUtc: Date
    /// Optional note.
    public let note: String?

    public init(kind: VitalKind, value: Double, atUtc: Date, note: String?) {
        self.kind = kind
        self.value = value
        self.atUtc = atUtc
        self.note = note
    }
}

/// A recorded allergy.
public struct Allergy: Sendable, Equatable, Codable {
    /// Stable identifier for the allergy.
    public let allergyId: String
    /// Substance the person is allergic to.
    public let substance: String
    /// Severity (e.g. "mild", "severe").
    public let severity: String

    public init(allergyId: String, substance: String, severity: String) {
        self.allergyId = allergyId
        self.substance = substance
        self.severity = severity
    }
}

/// A medication, optionally still active.
public struct Medication: Sendable, Equatable, Codable {
    /// Stable identifier for the medication.
    public let medId: String
    /// Medication name.
    public let name: String
    /// Dose (e.g. "10mg").
    public let dose: String
    /// Frequency (e.g. "daily").
    public let frequency: String
    /// When the medication was started (UTC).
    public let startedAtUtc: Date
    /// When the medication was stopped (UTC), or `nil` if still active.
    public let endedAtUtc: Date?

    public init(medId: String, name: String, dose: String, frequency: String,
                startedAtUtc: Date, endedAtUtc: Date?) {
        self.medId = medId
        self.name = name
        self.dose = dose
        self.frequency = frequency
        self.startedAtUtc = startedAtUtc
        self.endedAtUtc = endedAtUtc
    }
}

// MARK: - Errors

/// Errors thrown by the personal-health board.
public enum PersonalHealthError: Error, Equatable, CustomStringConvertible {
    /// `endMedication` referenced a med id that is not known.
    case unknownMedication(String)

    public var description: String {
        switch self {
        case .unknownMedication(let id): return "Unknown medication \(id)"
        }
    }
}

// MARK: - IPersonalHealthBoard

/// Vitals, allergies, and medications for the personal-health vertical.
/// User-scoped and never written to a shared store. A synchronous contract —
/// implementations are expected to be thread-safe.
public protocol IPersonalHealthBoard: AnyObject, Sendable {
    /// Records a vital reading.
    func record(_ v: VitalReading)
    /// Readings of `kind` at or after `since`, ascending by time.
    func readSince(kind: VitalKind, since: Date) -> [VitalReading]
    /// Most-recent reading of `kind`, or `nil`.
    func latest(kind: VitalKind) -> VitalReading?
    /// Adds (or replaces, by `allergyId`) an allergy.
    func addAllergy(_ a: Allergy)
    /// All recorded allergies.
    var allergies: [Allergy] { get }
    /// Adds (or replaces, by `medId`) a medication.
    func addMedication(_ m: Medication)
    /// Marks a medication ended. Throws when the med is unknown.
    func endMedication(medId: String, endedAtUtc: Date) throws
    /// Active medications (no end date), ordered by name.
    func activeMedications() -> [Medication]
}

// MARK: - InMemoryPersonalHealthBoard

/// Deterministic in-memory `IPersonalHealthBoard`. All state is guarded by a
/// single `NSLock`; every accessor returns an immutable snapshot.
public final class InMemoryPersonalHealthBoard: IPersonalHealthBoard, @unchecked Sendable {
    private let lock = NSLock()
    private var vitals: [VitalReading] = []
    private var allergiesById: [String: Allergy] = [:]
    private var meds: [String: Medication] = [:]

    public init() {}

    public func record(_ v: VitalReading) {
        lock.lock(); defer { lock.unlock() }
        vitals.append(v)
    }

    public func readSince(kind: VitalKind, since: Date) -> [VitalReading] {
        lock.lock(); defer { lock.unlock() }
        return vitals.filter { $0.kind == kind && $0.atUtc >= since }.sorted { $0.atUtc < $1.atUtc }
    }

    public func latest(kind: VitalKind) -> VitalReading? {
        lock.lock(); defer { lock.unlock() }
        return vitals.filter { $0.kind == kind }.max { $0.atUtc < $1.atUtc }
    }

    public func addAllergy(_ a: Allergy) {
        lock.lock(); defer { lock.unlock() }
        allergiesById[a.allergyId] = a
    }

    public var allergies: [Allergy] {
        lock.lock(); defer { lock.unlock() }
        return Array(allergiesById.values)
    }

    public func addMedication(_ m: Medication) {
        lock.lock(); defer { lock.unlock() }
        meds[m.medId] = m
    }

    public func endMedication(medId: String, endedAtUtc: Date) throws {
        lock.lock(); defer { lock.unlock() }
        guard let m = meds[medId] else { throw PersonalHealthError.unknownMedication(medId) }
        meds[medId] = Medication(medId: m.medId, name: m.name, dose: m.dose, frequency: m.frequency,
                                 startedAtUtc: m.startedAtUtc, endedAtUtc: endedAtUtc)
    }

    public func activeMedications() -> [Medication] {
        lock.lock(); defer { lock.unlock() }
        return meds.values.filter { $0.endedAtUtc == nil }.sorted { $0.name < $1.name }
    }
}

// MARK: - PersonalHealthDomainContext

/// Static domain-context constants for the personal-health vertical. Mirrors
/// `PersonalHealthDomainContext` in PersonalHealthDomainContext.cs.
public enum PersonalHealthDomainContext {
    /// System-prompt snippet injected ahead of user turns in this domain.
    public static let systemPromptSnippet = "[DOMAIN: Personal.Health] Personal health and wellness assistant. Help with symptom tracking, appointment preparation, medication reminders, health goal setting, nutrition basics, and health literacy. IMPORTANT: Always recommend consulting a qualified healthcare professional for medical decisions. This is not medical advice. Compliance: POPIA, Health Professions Act."
    /// Regulatory / compliance flags relevant to this domain.
    public static let complianceFlags: [String] = ["POPIA", "Health_Professions_Act", "Not_Medical_Advice"]
    /// Tools the Companion may suggest in this domain.
    public static let suggestedTools: [String] = ["health_tracker", "symptom_checker_ref", "calendar", "document_editor"]
}

File: kotlin/src/main/kotlin/com/bhengubv/circleai/personalhealth/PersonalHealth.kt
// PersonalHealth.kt
//
// Kotlin port of CircleAI.Personal.Health (PersonalHealthPrimitives.cs +
// PersonalHealthDomainContext.cs + PersonalHealthCompanionAdapter.cs) — the C#
// reference is the EXACT spec. Deterministic in-memory personal-health store:
// vitals, allergies, medications. Instances are user-scoped.
//
// Fidelity notes:
//   * C# `enum VitalKind` -> Kotlin `enum class` (same members, same order).
//   * C# `record` -> Kotlin `data class`.
//   * C# `double` (VitalReading.Value) -> Kotlin `Double`.
//   * C# `DateTimeOffset` -> `java.time.Instant`.
//   * C# `ConcurrentDictionary<string,_>` (Ordinal) -> `ConcurrentHashMap`;
//     vitals live in a plain list behind a lock.
//   * `ReadSince` keeps kind matches with AtUtc >= since, ordered ASC.
//   * `Latest` returns the newest reading for a kind, or null.
//   * `EndMedication` throws on an unknown medication.
//   * `ActiveMedications` returns meds with no end date, ordered by Name ASC.

package com.bhengubv.circleai.personalhealth

import com.bhengubv.circleai.companion.CompanionContext
import com.bhengubv.circleai.companion.CompanionProactiveEvent
import com.bhengubv.circleai.companion.CompanionTurn
import com.bhengubv.circleai.companion.ICompanionSession
import com.bhengubv.circleai.companion.InterfaceKind
import kotlinx.coroutines.flow.Flow
import java.time.Instant
import java.util.concurrent.ConcurrentHashMap

// =====================================================================
// Primitives (PersonalHealthPrimitives.cs)
// =====================================================================

/** The kind of a vital reading. Mirrors C# `VitalKind` (members + order preserved). */
enum class VitalKind {
    BloodPressureSystolic,
    BloodPressureDiastolic,
    GlucoseMgDl,
    WeightKg,
    HeartRateBpm,
    TemperatureC,
    OxygenPct,
    StepsCount,
}

/** A single vital measurement. Mirrors C# `VitalReading`. */
data class VitalReading(val kind: VitalKind, val value: Double, val atUtc: Instant, val note: String?)

/** A recorded allergy. Mirrors C# `Allergy`. */
data class Allergy(val allergyId: String, val substance: String, val severity: String)

/** A medication (optionally ended). Mirrors C# `Medication`. */
data class Medication(
    val medId: String,
    val name: String,
    val dose: String,
    val frequency: String,
    val startedAtUtc: Instant,
    val endedAtUtc: Instant?,
)

/** Deterministic personal-health board. Mirrors C# `IPersonalHealthBoard`. */
interface IPersonalHealthBoard {
    fun record(v: VitalReading)
    fun readSince(kind: VitalKind, since: Instant): List<VitalReading>
    fun latest(kind: VitalKind): VitalReading?
    fun addAllergy(a: Allergy)
    val allergies: List<Allergy>
    fun addMedication(m: Medication)
    fun endMedication(medId: String, endedAtUtc: Instant)
    fun activeMedications(): List<Medication>
}

/** In-memory [IPersonalHealthBoard]. Mirrors C# `InMemoryPersonalHealthBoard`. */
class InMemoryPersonalHealthBoard : IPersonalHealthBoard {
    private val vitals = mutableListOf<VitalReading>()
    private val allergies_ = ConcurrentHashMap<String, Allergy>()
    private val meds = ConcurrentHashMap<String, Medication>()
    private val lock = Any()

    override fun record(v: VitalReading) { synchronized(lock) { vitals.add(v) } }

    override fun readSince(kind: VitalKind, since: Instant): List<VitalReading> = synchronized(lock) {
        vitals.filter { it.kind == kind && it.atUtc >= since }.sortedBy { it.atUtc }
    }

    override fun latest(kind: VitalKind): VitalReading? = synchronized(lock) {
        vitals.filter { it.kind == kind }.maxByOrNull { it.atUtc }
    }

    override fun addAllergy(a: Allergy) { allergies_[a.allergyId] = a }

    override val allergies: List<Allergy>
        get() = allergies_.values.toList()

    override fun addMedication(m: Medication) { meds[m.medId] = m }

    override fun endMedication(medId: String, endedAtUtc: Instant) {
        val m = meds[medId] ?: throw IllegalStateException("Unknown medication $medId")
        meds[medId] = m.copy(endedAtUtc = endedAtUtc)
    }

    override fun activeMedications(): List<Medication> =
        meds.values.filter { it.endedAtUtc == null }.sortedBy { it.name }
}

// =====================================================================
// DomainContext (PersonalHealthDomainContext.cs)
// =====================================================================

/** Static domain context for Personal.Health. Mirrors C# `PersonalHealthDomainContext`. */
object PersonalHealthDomainContext {
    const val SYSTEM_PROMPT_SNIPPET: String =
        "[DOMAIN: Personal.Health] Personal health and wellness assistant. Help with symptom tracking, " +
            "appointment preparation, medication reminders, health goal setting, nutrition basics, and " +
            "health literacy. IMPORTANT: Always recommend consulting a qualified healthcare professional for " +
            "medical decisions. This is not medical advice. Compliance: POPIA, Health Professions Act."

    val complianceFlags: List<String> = listOf("POPIA", "Health_Professions_Act", "Not_Medical_Advice")

    val suggestedTools: List<String> =
        listOf("health_tracker", "symptom_checker_ref", "calendar", "document_editor")
}

// =====================================================================
// CompanionAdapter (PersonalHealthCompanionAdapter.cs)
// =====================================================================

/**
 * Wraps an [ICompanionSession] with the Personal.Health snippet + helpers.
 * Mirrors C# `PersonalHealthCompanionAdapter`.
 */
class PersonalHealthCompanionAdapter(private val inner: ICompanionSession) : ICompanionSession {
    override val sessionId: String get() = inner.sessionId
    override val identityId: String get() = inner.identityId
    override val interfaceKind: InterfaceKind get() = inner.interfaceKind
    override val history: List<CompanionTurn> get() = inner.history
    override val proactiveEvents: Flow<CompanionProactiveEvent> get() = inner.proactiveEvents

    override fun getContext(): CompanionContext = inner.getContext()
    override suspend fun refreshContextAsync() = inner.refreshContextAsync()
    override suspend fun signalFeedbackAsync(positive: Boolean, note: String?) =
        inner.signalFeedbackAsync(positive, note)
    override fun close() = inner.close()

    override suspend fun sendAsync(message: String): String = inner.sendAsync(enrich(message))
    override fun streamAsync(message: String): Flow<String> = inner.streamAsync(enrich(message))
    override suspend fun agentAsync(instruction: String): String = inner.agentAsync(enrich(instruction))

    private fun enrich(m: String): String = "${PersonalHealthDomainContext.SYSTEM_PROMPT_SNIPPET}\n\n$m"

    suspend fun prepareAppointmentAsync(symptoms: String, medHistory: String): String =
        inner.agentAsync("Help me prepare for a doctor appointment. Symptoms: $symptoms. Relevant history: $medHistory. Draft a concise symptom summary and list of questions to ask the doctor.")

    suspend fun explainHealthTermAsync(term: String): String =
        inner.agentAsync("Explain the medical term or concept in plain language: $term. Make it accessible to a non-medical person.")

    suspend fun interpretVitalsAsync(vitalsJson: String, age: String, baselineNotes: String): String =
        inner.agentAsync("Interpret vitals $vitalsJson for age $age. Baseline: $baselineNotes. Flag normal/borderline/concerning. Defer diagnosis to clinician.")

    suspend fun designSleepPlanAsync(currentPattern: String, targetWakeTime: String): String =
        inner.agentAsync("Design a sleep improvement plan from $currentPattern towards waking at $targetWakeTime. Cover light, caffeine, wind-down, environment.")

    suspend fun prepareForAppointmentAsync(concern: String, appointmentType: String): String =
        inner.agentAsync("Prepare for a $appointmentType about: $concern. Pre-visit checklist: symptoms log, questions, medication list, decisions to make.")

    suspend fun trackHabitImpactAsync(habit: String, vitalsBeforeAfter: String): String =
        inner.agentAsync("Analyse impact of $habit on vitals: $vitalsBeforeAfter. Confounders, signal strength, what to keep measuring.")
}
